Add a custom rejection handler for malformed requests

jisantuc opened the pull request:
## Overview

This PR transforms the messages associated with `MalformedRequestContentRejection`s from ugly and not user-friendly to pretty and user-friendly.

### Checklist

- [x] ~Styleguide updated, if necessary~
- [x] ~Swagger specification updated, if necessary~
- [x] ~Symlinks from new migrations present or corrected for any new migrations~

### Demo

If you're a cool kid with `httpie`, set an `RF_TOKEN` env variable, and:

```
http :9000/api/scenes/ Authorization: $RF_TOKEN foo=bar
http :9000/api/projects/ Authorization: $RF_TOKEN foo=bar
http :9000/api/thumbnails/ Authorization: $RF_TOKEN foo=bar
```

### Notes

##### 1

Setting up an implicit `RejectionHandler` where `bindAndHandle` is called is the [recommended way to implement custom rejection handling](http://doc.akka.io/docs/akka-http/10.0.1/scala/http/routing-dsl/rejections.html). I'm pretty sure we could do the same thing with the `UserErrorHandler` and not have to extend it in every `FooRoutes` trait.

If we do that, I think we throw both custom handlers into an `RfClientErrorHandler` as implicits and just import `RfClientErrorHandler._` in `Main.scala`. We'll get to delete more code! :smile: 

##### 2

I'm not sure how to list _all_ of the fields that are missing. Presumably people will be using our fancy and very informative docs site, so they should only make a few mistakes, but I can imagine going a little insane trying to brute force my way to a legal request, and we don't list all the fields right now anyway.

## Testing Instructions

 * `./scripts/server` (or `./scripts/console api-server ./sbt`, then `api/run` if you're impatient)
 * Send some bad json to any endpoint you choose
 * Verify that the error message tells you what's wrong instead of mysteriously complaining about decoding and `DownField(something-or-other)`s

Closes #1398

ddohler commented:
More info on the UUID parsing behavior for later diagnosis.

POSTING to `/api/uploads/`
Correct, should work if you have this organization and datasource in your DB:
```json
{
"organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
  "uploadStatus": "UPLOADING",
  "fileType": "GEOTIFF",
  "uploadType": "LOCAL",
  "files": [],
  "datasource": "697a0b91-b7a8-446e-842c-97cda155554d",
  "metadata": {},
  "visibility":"PRIVATE"
}
```

Incorrect (truncated the three middle fields of the `datasource` UUID by one character):
```json
{
"organizationId": "dfac6307-b5ef-43f7-beda-b9f208bb7726",
  "uploadStatus": "UPLOADING",
  "fileType": "GEOTIFF",
  "uploadType": "LOCAL",
  "files": [],
  "datasource": "697a0b91-7a8-46e-42c-97cda155554d",
  "metadata": {},
  "visibility":"PRIVATE"
}
```

Returns a `501` and results in this console error:
```
postgres_1     | ERROR:  insert or update on table "uploads" violates foreign key constraint "uploads_datasource_fkey"
postgres_1     | DETAIL:  Key (datasource)=(697a0b91-07a8-046e-042c-97cda155554d) is not present in table "datasources".
```
Note that each of the truncated UUID fields has been left-padded with a 0.

jisantuc commented:
Bad types do have a slightly different format -- you don't get the "Attempt to decode value..." before the cursor error. I added parsing for that and cleaned up the pattern match to use extractors with my compiled regex instead of matching over `findFirstMatchIn`, so each compiled regex now fully describes its string.

ddohler commented:
:+1: Tested the changes, they work great!
